### PR TITLE
protocol contract deployment verification via CI

### LINKED ISSUE #1594: Add protocol contract deployment verification to CI
All protocol contracts deployments can now be verified locally https://github.com/api3dao/airnode/pull/1590 We should add deployment verification to CI similar to manager-multisig.

### PR DESCRIPTION
**Pull request by Ashar2shahid**
Closes #1594 

This add CI Verification to the protocol contracts using the `verify-local` script. I opted to return the list of all failed verifications and then throw the error instead of throwing an error at the first failed verification. 

- I created a separate workflow file for deployment-verification due to the [path filter](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore) only being applicable on a per workflow basis, this means that if we want to run the CI only if the `airnode-protocol` folder changes then we need a separate workflow.
 - we can use the [paths-filer action](https://github.com/dorny/paths-filter) if we want to do the deployment-verification and check if the `airnode-protocol` folder has changed in the `Continuous Build` workflow but its a third-party.
 - we can opt out of checking if the `airnode-protocol` folder has changed and run the CI on every workflow-run by incorporating it as an extra step of the `Continuous Build` workflow

### CODE REVIEW

**Comment by Ashar2shahid** [comment]
As seen here: https://github.com/api3dao/airnode/actions/runs/4041541899/jobs/6948239143 the CI is failing because of some chains not existing (roptsen,rinkeby,kovan), going down (gnosis-testnet) frequently or the RPCs being faulty(polygon)

File: packages/airnode-protocol/scripts/verify-local.ts [hunk 1]
```diff
@@ -4,73 +4,134 @@
...
+          totalTimeoutMs: 50000,
+          delay: {
+            type: 'random',
+            minDelayMs: 2000,
+            maxDelayMs: 5000,
+          },
+        });
+        if (!goFetchCreationTx.success || !goFetchCreationTx.data) {
+          return {
+            status: false,
```
Comment (bbenligiray) [inline packages/airnode-protocol/scripts/verify-local.ts:86]:
```suggestion
            success false,
```

**Review by bbenligiray — COMMENTED** [review]
I took a brief look, I can review again after https://github.com/api3dao/airnode/pull/1562 is merged

Just remove the ropsten, rinkeby, kovan deployments. It should have been done as a part of https://github.com/api3dao/airnode/issues/1280 While at it, remove `gnosis-testnet` and `metis-testnet` too. They can be added back in a separate PR to close https://github.com/api3dao/airnode/issues/1628.

File: .github/workflows/protocol-verify.yml [hunk 2]
```diff
@@ -0,0 +1,93 @@
+name: Verify Contract Deployments
+
+on:
+  push:
+  pull_request:
+    branches:
+      - master
+    paths:
+      - 'packages/airnode-protocol/**'
```
[the remark this answers is not in the document]
Comment (Ashar2shahid) [reply .github/workflows/protocol-verify.yml:9]: if the protocol contracts change by mistake, then I think we should be running the CI to make sure the localbyte code matches the one onchain.

File: packages/airnode-protocol/package.json [hunk 3]
```diff
@@ -19,10 +19,10 @@
     "deploy:deterministic": "DETERMINISTIC=true hardhat deploy --network $NETWORK --tags deploy && yarn run deploy:generate-references",
     "deploy:undeterministic": "hardhat deploy --network $NETWORK --tags deploy && yarn run deploy:generate-references",
     "deploy:verify": "hardhat deploy --network $NETWORK --tags verify",
-    "deploy:verify-local": "hardhat run scripts/verify-local.ts --network $NETWORK",
```
[the remark this answers is not in the document]
Comment (Ashar2shahid) [reply packages/airnode-protocol/package.json:22]: cool I'll try to accommodate both

[the remark this answers is not in the document]
Comment (Ashar2shahid) [reply packages/airnode-protocol/package.json:22]: use `test:verify-local:network` for single network verification

File: packages/airnode-protocol/scripts/verify-local.ts [hunk 4]
```diff
@@ -4,73 +4,134 @@
...
+          };
+        } else {
+          const salt = hre.ethers.constants.HashZero;
+          const deterministicDeploymentAddress = hre.ethers.utils.getCreate2Address(
+            '0x4e59b44847b379578588920ca78fbf26c0b4956c', // default create2 factory address in hardhat
+            salt,
+            hre.ethers.utils.keccak256(generatedBytecode)
+          );
+          generatedBytecode = hre.ethers.utils.hexConcat([salt, generatedBytecode]);
+          assert(deterministicDeploymentAddress === deployment.address);
```
[the remark this answers is not in the document]
Comment (Ashar2shahid) [reply packages/airnode-protocol/scripts/verify-local.ts]: moving it into `verifyDeployedBytecode()` made the code too ugly, I moved the assertion below

File: .github/workflows/protocol-verify.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (bbenligiray) [reply .github/workflows/protocol-verify.yml:9]: Hmm I suppose so

**Comment by bbenligiray** [comment]
Can you merge https://github.com/api3dao/airnode/pull/1562 and merge master to 1594 so that we know that the CI also works for the new chains?

**Comment by Ashar2shahid** [comment]
Checks are passing, good to merge

**Review by bbenligiray — APPROVED** [review]
I think the implementation in `verify-local.ts` could be better compartmentalized to be more reusable. The current thing addresses the issue though so maybe someone will come back to this later.

**Pull request merged**